composer/utils/object_store/sftp_object_store.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Utility for uploading to and downloading from cloud object stores."""
+import os
+import pathlib
+import urllib.parse
+import uuid
+from typing import Any, Callable, Dict, Optional, Union
+
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore, ObjectStoreTransientError
+
+__all__ = ['SFTPObjectStore']
+
+try:
+    from paramiko import SSHClient
+    _PARAMIKO_AVAILABLE = True
+except ImportError:
+    _PARAMIKO_AVAILABLE = False
+
+
+def _set_kwarg(value: Any, kwargs: Dict[str, Any], arg_name: str, kwarg_name: str):
+    if kwarg_name in kwargs:
+        raise ValueError(f'The `{arg_name}` should be not be specified directly if also included via `connect_kwargs`')
+    kwargs[kwarg_name] = value
+
+
+def _ensure_transient_errors_are_wrapped(e: Exception):
+    from paramiko import SSHException
+    if isinstance(e, SSHException):
+        if 'Server connection dropped:' in str(e):
+            raise ObjectStoreTransientError from e
+    raise e
+
+
+class SFTPObjectStore(ObjectStore):
+    """Utility for uploading to and downloading to a server via SFTP.
+
+    Args:
+        host (str): The server to connect to.
+
+            Also accepts a URI string in the form ``'sftp://username@host:port/./relative/path'``.
+            For an absolute path, use a double `//` -- e.g. ``'sftp://username@host:port//absolute/path'``.
+
+        port (int, optional): The server port to connect to.
+        username (str, optional): The username (if not specified in the SSH config) needed to authenticate.
+            Defaults to None.
+        password (str, optional): The password (if required) needed to authenticate. Defaults to None.
+        known_hosts_filename (pathlib.Path | str, optional): The filename of the private key.
+        cwd (str, optional): The directory to navigate to upon creating the SSH connection. If not present
+            it will be created.
+        connect_kwargs (Dict[str, Any], optional): Any additional kwargs to pass through to :meth:`.SSHClient.connect`.
+    """
+
+    def __init__(
+        self,
+        host: str,
+        port: int = 22,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        known_hosts_filename: Optional[Union[pathlib.Path, str]] = None,
+        cwd: Optional[str] = None,
+        connect_kwargs: Optional[Dict[str, Any]] = None,
```

Comment:
The private key will need to be passed via `connect_kwargs` because we cannot rely on only one SSH key (`id_rsa`) being available. Would prefer a top level argument like `private_key_path`. Then you'll need:

```python
pkey = RSAKey.from_private_key_file(private_key_path)
if private_key_path:
            _set_kwarg(pkey, connect_kwargs, arg_name='private_key_path', kwarg_name='pkey')
```

Reply:
Agreed; just added that in!